File: _algos/strings/trie/tst.py
#!python3

"""
Ternary Search Trie
"""

class Node:
    def __init__(self, ch):
        self.left = None
        self.right = None
        self.middle = None
        self.value = None
        self.ch = ch

class TST:
    def __init__(self, R):
        self.nodes = [None] * R
        self.R = R

    def _char_at(self, key, i):
        ch = ord(str.lower(key[0])) - ord('a')
        ch %= self.R
        return ch

    def _put(self, node, key, i, value):
        if node == None:
            node = Node(key[i])
        if i == len(key) - 1:
            node.value = value
            return node
        elif key[i] < node.ch:
            node.left = self._put(node.left, key, i, value)
        elif key[i] > node.ch:
            node.right = self._put(node.right, key, i, value)
        else:  # ch == key[i]
            node.middle = self._put(node.middle, key, i+1, value)
        return node

    def put(self, key, value):
        if not key:
            return
        ch = self._char_at(key, 0)
        self.nodes[ch] = self._put(self.nodes[ch], key, 0, value)

    def search(self, key):
        ch = self._char_at(key, 0)
        return self._search(self.nodes[ch], key, 0)

    def _search(self, node, key, i):
        if node != None:
            if i == len(key) - 1:
                return node.value
            elif key[i] < node.ch:
                return self._search(node.left, key, i)
            elif key[i] > node.ch:
                return self._search(node.right, key, i)
            else:  # ch == key[i]
                return self._search(node.middle, key, i+1)
        return None

    def _inorder(self, arr, node, s):
        if node == None:
            return
        self._inorder(arr, node.left, s)
        if node.value:
            arr += [(''.join(s + [node.ch]), node.value)]
        self._inorder(arr, node.middle, s + [node.ch])
        self._inorder(arr, node.right, s)

    def inorder(self, arr):
        for i,e in enumerate(self.nodes):
            self._inorder(arr, e, [])

    def print(self):
        arr = []
        self.inorder(arr)
        print(arr)


if __name__ == '__main__':
    radix = ord('z') - ord('a') + 1
    tst = TST(radix)
    tst.put('she', 1)
    assert(tst.search('she') == 1)
    tst.put('shell', 2)
    assert(tst.search('shell') == 2)
    assert(tst.search('shel') == None)
    tst.put('abc', 3)
    assert(tst.search('a') == None)
    assert(tst.search('ab') == None)
    assert(tst.search('abc') == 3)
    tst.print()

File: _algos/strings/trie/tst.c
//
// Author: Artyom K. <artyomkarpov [at] gmail [dot] com>
//
// Ternary search trie
//  Time: O(log(N))
//  Space: O(NW)
//
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>

typedef struct node_s {
    struct node_s * left;
    struct node_s * right;
    struct node_s * middle;
    int val;
    char c;
} Node;

Node * put(Node * n, const char * key, int keylen, int d);

int search(Node * n, const char * key, int keylen, int d);

void printtrie(Node *n, int level);

int main()
{
    int n = 4;
    const char * strings[n];
    strings[0] = "abcd";
    strings[1] = "def";
    strings[2] = "fghi";
    strings[3] = "ijklm";
    Node * root = NULL;
    int i;
    for (i = 0; i < n; ++i) {
        root = put(root, strings[i], strlen(strings[i]), 0);
    }

    for (i = 0; i < n; ++i) {
        assert(search(root, strings[i], strlen(strings[i]), 0) == 1);
    }
    assert(search(root, "not present", 11, 0) == 0);
    assert(search(root, "another not present", 19, 0) == 0);

    printtrie(root, 0);

    printf("done\n");
    return 0;
}

Node * put(Node * n, const char * key, int keylen, int d)
{
    if (n == NULL)
    {
        n = (Node *) malloc(sizeof(Node));
        n->left = NULL;
        n->right = NULL;
        n->middle = NULL;
        n->val = 0;
        n->c = key[d];
    }
    if (n->c == key[d])
    {
        if (keylen - 1 == d)
            n->val = 1;
        else
            n->middle = put(n->middle, key, keylen, d+1);
    }
    else if (key[d] < n->c)
        n->left = put(n->left, key, keylen, d);
    else
        n->right = put(n->right, key, keylen, d);

    return n;
}

int search(Node * n, const char * key, int keylen, int d)
{
    if (n == NULL)
        return 0;

    if (n->c == key[d])
    {
        if (keylen - 1 == d)
            return n->val;
        else
            return search(n->middle, key, keylen, d+1);
    }
    else if (key[d] < n->c)
        return search(n->left, key, keylen, d);
    else 
        return search(n->right, key, keylen, d);
}

void printtrie(Node * n, int level)
{
    if (n == NULL)
        return;
    int i;
    for (i = 0; i < level; ++i) 
        printf(" ");

    printf("ch='%c' val=%d\n", n->c, n->val);
    printtrie(n->left, level + 1);
    printtrie(n->middle, level + 1);
    printtrie(n->right, level + 1);
}
